### ISSUE TITLE
FR: have remote_branches() take the remote name as an argument

### ISSUE THREAD

**Issue by avamsi** [martinvonz/jj#1011]
For example, I have the following setup with the jj repo --

```
$ jj git remote list

fork https://github.com/avamsi/jj.git
origin https://github.com/martinvonz/jj.git
```

(And a similar setup at work as we don't have dev branches on the main repo itself.)

This will be useful in limiting `jj log` to changes not in the `origin` remote but in the `fork` remote.

**Comment by martinvonz** [comment]
Make sense. That shouldn't be hard to add in https://github.com/martinvonz/jj/blob/main/lib/src/revset.rs if you're interested.

**Comment by avamsi** [comment]
I have a draft at #1013, PTAL -- I filed the FR for a single argument and ended up implementing for multiple arguments, just wanted to run that by you before editing any docs / changelog e.t.c.

Also, the current behavior is to just return an empty set when an unknown remote is provided but I'm not sure if it would be better to throw an error and let the user be explicit about presence (like with `present()` or something).

**Comment by yuja** [comment]
I think 0 or 1 argument is slightly better. If `remote_branches(a, b, c)` meant `a || b || c`, `remote_branches()` could be considered "no match". If it takes 0 or 1 argument, maybe we can say the default parameter is "match all".

fwiw, we might want a revset to select local branches by name. If it were `branches("push-")` (to select `*push-*`), `remote_branches()` taking a remote name (not branch name) might be confusing. @martinvonz?

**Comment by martinvonz** [comment]
Those are both good points.

I agree that it's simple enough for the user to say `remote_branches(a) | remote_branches(b) | remote_branches(c)` to match all of them.

We don't have keyword arguments like hg does, so we can't easily make syntax be `remote_branches(remote="origin")`. I'm not sure what to do instead. We could perhaps have the user write `remote_branches(@origin)`. I'm not sure how to extend that to let the user say "branches containing 'foo' on remote 'bar'". Another option is to have different functions for matching branches on a remote, but I don't know what to call that since the name `remote_branches()` is taken :)

**Comment by martinvonz** [comment]
I suppose we could let the user say `remote_branches("", foo)` to match all branches containing `foo` on any remote, and `remote_branches(origin[, ""])` to match all branches on "origin", but maybe it's too weird to require an explicit empty string to match all remotes for the first case.

**Comment by avamsi** [comment]
Maybe allow them to be regular expressions with the signatures like `branches(branch_name_regex)` and `remote_branches(remote_name_regex, branch_name_regex)`?

Above examples become `branches(push-.+)`, `remote_branches(.+, foo)`, `remote_branches(origin, .+)`, that doesn't really solve the empty string problem but at least it makes it more explicit, IMO (and the arguments could default to `.+`).

**Comment by martinvonz** [comment]
Sounds good in general, but I think we should make sure that all "needles" (things to match against) in revset expressions use the same syntax if possible. So far, I think we have e.g. `description()` and `author()`, which match the needle as a substring. So I'd say we should stick with that for `remote_branches()`, and then later extend it to allow e.g. `re:push-.*` or `literal:foo.bar`.

I'm fine with having the user specify that empty string for now.

@yuja, what do you think? I realize that this proposal doesn't address your concern that `branches(foo)` refers to branches while `remote_branches(foo)` refers to remotes. Do you have a better suggestion?

**Comment by avamsi** [comment]
Makes sense. Could you clarify the API signature you're thinking about?

> I'm fine with having the user specify that empty string for now.

Does this mean you don't want to make the arguments optional (and let this be a breaking change)?
>
> @yuja, what do you think? I realize that this proposal doesn't address your concern that branches(foo) refers to branches while remote_branches(foo) refers to remotes. Do you have a better suggestion?

Why not just `branches(branch_name_substr)` and `remote_branches(remote_name_substr, branch_name_substr)`?

**Comment by martinvonz** [comment]
> Does this mean you don't want to make the arguments optional (and let this be a breaking change)?

Sorry, that was a bit unclear. I just meant that if the user wants to match branches named "foo" on all remotes, it seems fine to have them pass an empty string for the remote name. Unless we think of something better, of course.

> Why not just `branches(branch_name_substr)` and `remote_branches(remote_name_substr, branch_name_substr)`?

It's a weird for the meaning of the first argument to change depending on how many arguments you pass. It would also mean that you'd have to do `remote_branches("", origin)` for your usecase, right?

**Comment by PhilipMetzger** [comment]
I think we should implement this feature for both `branches()` and `remote_branches()` so they're symmetrical. 
This would also fix #252.
On a side note, we had a related discussion in Discord, on the topic of release branches,  which begins [here](https://discord.com/channels/968932220549103686/969829516539228222/1049019598235770960).

**Comment by yuja** [comment]
> I'm fine with having the user specify that empty string for now.

Yep. I considered `remote_branches("branch@remote")`, but user would have to write regex or glob for substring match. So maybe `remote_branches([branch[, remote]])`/`branches([branch])`, and `remote_branches("", remote)` to select all branches from remote? If we add keyword arguments, it can be expressed better.

**Comment by martinvonz** [comment]
> So maybe `remote_branches([branch[, remote]])`/`branches([branch])`, and `remote_branches("", remote)` to select all branches from remote?

Sounds good. What do you think, @avamsi?

**Comment by avamsi** [comment]
SGTM, I'll update the PR.

**Comment by avamsi** [comment]
Updated the code in #1013, PTAL (I just want to make sure the code is more or less fine before making any doc changes).

@PhilipMetzger, thanks for linking #252, I didn't realize that existed -- #1013 in it's current state should fix both of these, I think.

**Comment by martinvonz** [comment]
The code looks good, but I don't think it actually fixes #252 because the `jj branches` command works by iterating over branches, so filtering by a revset doesn't help it.

**Comment by avamsi** [comment]
Forgot to close this (was fixed by #1013). Re: #252, yeah, sorry, misread it.
